### PR TITLE
Integration test for "category deletion and the impact on associated products"

### LINKED ISSUE #228: Category deletetion and impact on associated products
(Special thanks to @marionf, @matks, @Julievrz who already contributed to this VERY important matter.
[The issue](https://github.com/PrestaShop/PrestaShop/issues/9811) hugely affects how products are browsed by categories in FO).

### **What happens to the default category of a product P associated to category C when C is deleted?**

**Case 1 : Product P has only category C (which also means C is default category)**
When merchant deletes category and chooses one of the two first options in modal (doesn't choose to delete associated products)  THEN associate P to the parent category and set parent category as default.
case 1

**Case 2 : Product P has Category C and other categories**
***Case 2.1: Category C is default category*** 
THEN default category becomes parent of C
case 2 1

***Case 2.2 :  Category C is not default category***
THEN default category is kept unchanged.
case 2 2

Behavior tests here: https://github.com/PrestaShop/PrestaShop/pull/24668

### PR DESCRIPTION
**Pull request by arouiadib**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      |  Integration test for category deletion and impact on associated products.
| Type?             | refacto 
| Category?         | TE 
| BC breaks?        |  no
| Deprecations?     |  no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/prestashop-specs/issues/228
| How to test?      | run `php ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml --name="Delete category"`
| Possible impacts? | None

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24668)

### CODE REVIEW

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 1]
```diff
@@ -35,23 +35,35 @@
  */
 abstract class AbstractDeleteCategoryHandler
 {
+    /**
+     * Get products that belong to a category
+     *
+     * @param int $categoryId
+     */
+    protected function getProductsWithoutCategory($categoryId)
```
Comment (kpodemski) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]: @arouiadib I think you meant `FromCategory`?

Comment (PululuK) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]: Products without category or category without products ?
I feel like the name of the function doesn't really match what it does

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 2]
```diff
@@ -35,23 +35,35 @@
...
+    {
+        $productsWithoutCategory = \Db::getInstance()->executeS('
+                    SELECT p.`id_product`
+                    FROM `' . _DB_PREFIX_ . 'product` p
+                    ' . Shop::addSqlAssociation('product', 'p') . '
+                    WHERE EXISTS (
+                        SELECT 1 FROM `' . _DB_PREFIX_ . 'category_product` cp
+                        WHERE cp.`id_product` = p.`id_product`
+                        AND cp.`id_category` = ' . $categoryId . '
```
Comment (kpodemski) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]:
```suggestion
                        AND cp.`id_category` = ' . (int) $categoryId . '
```

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php (hunk 1 above)
Comment (kpodemski) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]:
```suggestion
    protected function getProductsWithoutCategory(int $categoryId): array
```

make sure to return empty array here if you don't have any results

Comment (kpodemski) [reply src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]: I think this query belongs to the `Category.php`, not to handler

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 3]
```diff
@@ -35,23 +35,35 @@
...
      * @param CategoryDeleteMode $mode
      */
-    protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $mode)
+    protected function handleProductsUpdate($productsWithoutCategory, $parentCategoryId, CategoryDeleteMode $mode)
     {
-        $productsWithoutCategory = \Db::getInstance()->executeS('
-			SELECT p.`id_product`
-			FROM `' . _DB_PREFIX_ . 'product` p
-			' . Shop::addSqlAssociation('product', 'p') . '
-			WHERE NOT EXISTS (
```
Comment (kpodemski) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]: And here there was `not exists` 🤔

**Review by kpodemski — CHANGES_REQUESTED** [review]

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php (hunk 1 above)
Comment (arouiadib) [reply src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]: Hi @kpodemski , @PululuK 
Thanks for CR, I will take them into account but not in this PR. 
Sorry I made mistake of pushing Fix+Test while the aim of this PR is only testing.
My goal is to provide with some tests that fails and have them green with the fix. I have separate PR for the fix where I will put your remarks.
:heart:

**Comment by arouiadib** [comment]
@kpodemski @PululuK  Thanks a lot for your first feedback.
The PR now contains only tests and I would love to receive your CR.

The changes you already suggested will be taken into account in another PR (that contains the fix).

**Comment by PierreRambaud** [comment]
Hi @arouiadib are you still working on it because CI is failing?

**Pull request closed without merging**